src/data/ellpack_page_source.cu

```diff
@@ -0,0 +1,184 @@
+            dmlc::SeekStream* fi,
+            const std::vector<bst_uint>& sorted_index_set) override {
+    auto* impl = page->Impl();
+    if (!fi->Read(&impl->n_rows))  return false;
+    return fi->Read(&page->Impl()->idx_buffer);
+  }
+
+  void Write(const EllpackPage& page, dmlc::Stream* fo) override {
+    auto* impl = page.Impl();
+    fo->Write(impl->n_rows);
+    auto buffer = impl->idx_buffer;
+    CHECK(!buffer.empty());
+    fo->Write(buffer);
+  }
+};
+
+// Build the quantile sketch across the whole input data, then use the histogram cuts to compress
+// each CSR page, and write the accumulated ELLPACK pages to disk.
+EllpackPageSourceImpl::EllpackPageSourceImpl(DMatrix* dmat,
+                                             const std::string& cache_info,
+                                             const BatchParam& param) noexcept(false) {
+  device_ = param.gpu_id;
+
+  monitor_.Init("ellpack_page_source");
+  dh::safe_cuda(cudaSetDevice(device_));
+
+  monitor_.StartCuda("Quantiles");
+  common::HistogramCuts hmat;
+  size_t row_stride =
+      common::DeviceSketch(device_, param.max_bin, param.gpu_batch_nrows, dmat, &hmat);
+  monitor_.StopCuda("Quantiles");
+
+  monitor_.StartCuda("CreateEllpackInfo");
+  ellpack_info_ = EllpackInfo(device_, dmat->IsDense(), row_stride, hmat, ba_);
+  monitor_.StopCuda("CreateEllpackInfo");
+
+  monitor_.StartCuda("WriteEllpackPages");
+  WriteEllpackPages(dmat, cache_info);
+  monitor_.StopCuda("WriteEllpackPages");
+
+  source_.reset(new SparsePageSource<EllpackPage>(cache_info, kPageType_));
+}
+
+void EllpackPageSourceImpl::BeforeFirst() {
+  source_->BeforeFirst();
+}
+
+bool EllpackPageSourceImpl::Next() {
+  return source_->Next();
+}
+
+EllpackPage& EllpackPageSourceImpl::Value() {
+  EllpackPage& page = source_->Value();
+  page.Impl()->InitDevice(device_, ellpack_info_);
+  return page;
+}
+
+const EllpackPage& EllpackPageSourceImpl::Value() const {
+  EllpackPage& page = source_->Value();
+  page.Impl()->InitDevice(device_, ellpack_info_);
+  return page;
+}
+
+// Compress each CSR page to ELLPACK, and write the accumulated pages to disk.
+void EllpackPageSourceImpl::WriteEllpackPages(DMatrix* dmat, const std::string& cache_info) const {
+  auto cinfo = ParseCacheInfo(cache_info, kPageType_);
+  SparsePageWriter<EllpackPage> writer(cinfo.name_shards, cinfo.format_shards, 6);
+  std::shared_ptr<EllpackPage> page;
+  writer.Alloc(&page);
+  auto* impl = page->Impl();
+  impl->matrix.info = ellpack_info_;
+  impl->Clear();
+
+  const MetaInfo& info = dmat->Info();
+  size_t bytes_write = 0;
+  double tstart = dmlc::GetTime();
+  for (const auto& batch : dmat->GetBatches<SparsePage>()) {
+    impl->Push(device_, batch);
+
+    if (impl->MemCostBytes() >= DMatrix::kPageSize) {
+      bytes_write += impl->MemCostBytes();
+      writer.PushWrite(std::move(page));
+      writer.Alloc(&page);
+      impl = page->Impl();
+      impl->matrix.info = ellpack_info_;
```

Comment:
is the ellpack info also written with the compressed data for every page? if a dataset contains large number of features, wouldn't this cost be prohibitive across a large number of such pages?

Reply:
> is the ellpack info also written with the compressed data for every page?

@sriramch I don't think it's written to disk for every page according to `EllpackPageRawFormat::Write`.  @rongou Correct me if i'm wrong.

Reply:
Right, it's kept in memory.